src/test/java/com/exactpro/th2/taskutils/FutureTrackerTest.java

```diff
@@ -63,104 +104,334 @@ private CompletableFuture<Integer> chainFuture (CompletableFuture<Integer> futur
 
             return res;
         });
+        futuresList.add(nextFuture);
+        return nextFuture;
     }
 
-    @Test
-    public void testEmptyTracker () {
-        long expectedTrackingMillis = NO_DELAY_MILLIS;
+    @SuppressWarnings("SameParameterValue")
+    private void assertLessDuration(long timeoutMillis, Task task) throws InterruptedException {
+        long start = System.nanoTime();
+        task.invoke();
+        long actualTrackingMillis = (System.nanoTime() - start) / 1_000_000;
+        assertThat(actualTrackingMillis).isLessThan(timeoutMillis);
+    }
 
-        FutureTracker<Integer> futureTracker = new FutureTracker<>();
+    @SuppressWarnings("SameParameterValue")
+    private void assertGreaterDuration(long timeoutMillis, Task task) throws InterruptedException {
         long start = System.nanoTime();
-        futureTracker.awaitRemaining();
+        task.invoke();
+        long actualTrackingMillis = (System.nanoTime() - start) / 1_000_000;
+        assertThat(actualTrackingMillis).isGreaterThanOrEqualTo(timeoutMillis);
+    }
 
-        Assertions.assertThat(futureTracker.isEmpty()).isTrue();
+    @AfterAll
+    public void afterAll() throws InterruptedException {
+        executor.shutdown();
+        if (!executor.awaitTermination(5, SECONDS)) {
+            executor.shutdownNow();
+        }
+    }
 
-        long actualTrackingMillis = (System.nanoTime() - start)/1_000_000;
-        Assertions.assertThat(actualTrackingMillis).isLessThan(expectedTrackingMillis);
+    @AfterEach
+    public void afterEach() throws InterruptedException {
+        for (StartableRunnable startableRunnable : runnableList) {
+            if (!startableRunnable.isReady()) {
+                startableRunnable.awaitReadiness();
+            }
+            if (!startableRunnable.isStarted()) {
+                startableRunnable.start();
+            }
+        }
+        runnableList.clear();
+        for (CompletableFuture<?> future : futuresList) {
+            try {
+                future.cancel(true);
+                future.get();
+            } catch (InterruptedException e) {
+                throw e;
+            } catch (Exception e) {
+                // do nothing
+            }
+        }
     }
 
-    @Test
-    public void testTrackingSingleFuture () {
-        long expectedTrackingMillis = DELAY_MILLIS;
+    @Nested
+    class Both {
```

Comment:
question: What does `both` mean here? If it means `Unlimited and with limit` we probably should write it in the name of the class and not hide behind a single word